Refactor-Navbar-Themes

Alixers505 opened the pull request:
Theme Sass Refactoring 
-- Phase 1: Navbar
---- Stage 1: Navbar Link Colors

**_variables:**
- Removed unnecessary variable settings
- Added variable folder, with _navbar subfile
- Moved navbar settings into _navbar subfile
- Added import for _navbar subfile
- Created nested color map for navbar themes (currently only contains link settings, per Stage 1 objective)

**_functions:** 
- Updated map function to be usable across multiple map types, including nested maps
- Removed old navbar link function

**_mixins**
- Updated theme-navbar-link-color to utilize new function and theme map

**themes/_navbar**
- Updated mixin calls to utilize new format, referencing new map and theme key

**bizbash _directory**
- Updated theme-navbar-link-color to reference new map and theme key

**bizbash index**
- Added theme map for bizbash custom theme
- Updated theme-navbar-link-color references to bizbash map and theme keys
- Removed unused variables
- Updated #fff references to use $white

ToDo: reduce bizbash line additions?

Next Stage -- Stage 2: Navbar Decorations & Backgrounds

Alixers505 commented:
Not sure what I can do for this build error:
https://travis-ci.org/endeavorb2b/websites/builds/566756618?utm_source=github_status&utm_medium=notification

zarathustra323 commented:
> Not sure what I can do for this build error:
> https://travis-ci.org/endeavorb2b/websites/builds/566756618?utm_source=github_status&utm_medium=notification

I'm re-running the build

zarathustra323 commented on packages/themes/default/styles/variables/_navbar.scss:
```diff
@@ -0,0 +1,115 @@
+    hover: $theme-site-navbar-dark-hover-color,
+    active: $theme-site-navbar-dark-active-color
+  ),
+  light: (
```
How would I override these colors now, to something other than the pre-defined dark/light values? The reason there were variables for each navbar  type is so they could be individually overridden (if need be) regardless of what type was selected

Alixers505 replied:
From how I am reading this, it would be the same way you would have done it before, as they are still pulling from variable settings. I may be misunderstanding your question however. If I am misunderstanding, could you provide an example of the kind of change of which you are speaking?

With this change, there are still different navbar types, they have just been consolidated to dark and light, as the primary, secondary, and tertiary variables were all using the same default dark settings.

zarathustra323 commented on sites/bizbash/server/styles/index.scss:
```diff
@@ -19,8 +19,27 @@ $black:     #000;
+    hover: $white,
+    active: $white
+  ),
+  lighter: (
```
This doesn't follow the stand convention for this variable... if anything, the root var should support merging additional values from an extending site

Alixers505 replied:
This was a stopping point for last night, hence the addition of a "ToDo" in my initial merge comment.

zarathustra323 commented on sites/bizbash/server/styles/index.scss:
```diff
@@ -246,6 +255,20 @@ body {
   }
+  &--secondary {
+    #{ $self }__link {
+      @include theme-navbar-link-color($theme-site-bizbash-map, light);
```
The root theme should handle this

Alixers505 replied:
This one needs to be there, since it is using light instead of dark, but both tertiary and primary includes can be removed, since they are using dark (which is what the default is). Thanks! Likely stared too long at the same files last night!

Alixers505 replied:
Oh, and I did update it to use the theme-site-navbar-map, rather than a separate map on bizbash's file.

zarathustra323 commented on sites/bizbash/server/styles/index.scss:
```diff
@@ -246,6 +255,20 @@ body {
+  }
+  &--tertiary {
+    #{ $self }__link {
+      @include theme-navbar-link-color($theme-site-bizbash-map, dark);
```
The root theme should handle this

Alixers505 commented on packages/themes/default/styles/variables/_navbar.scss:
```diff
@@ -0,0 +1,115 @@
+$theme-site-navbar-light-active-color:  $theme-site-navbar-light-hover-color !default;
+
+// Themes
+$theme-site-navbar-map: (
```
The map will later be used for more than just colors, hence the use of a more broad terminology at this time.

zarathustra323 replied:
I'm a little worried about using such a deep map for all of these values. Why not maintain individual variables like before?

zarathustra323 replied:
Also, the mixin is now dependent on the map variable and its structure. Previously, the mixin could be use with _any_ color values (like BizBash was doing)

Alixers505 replied:
Yes, the mixin is now dependent upon having a map passed in; however, the functionality of it has not changed, beyond the fact that it can now support nested maps--which reduces the likelihood of duplicate code as it keeps the definitions for a component all housed within the same area. This will become especially useful in future stages, as the same mixin can be utilized to pull in more than just link color assignments.

Here's an example of where this map can lead:

```
// Colors

// Dark
$theme-site-navbar-dark-color:          rgba($white, .75) !default;
$theme-site-navbar-dark-hover-color:    $white !default;
$theme-site-navbar-dark-active-color:   $theme-site-navbar-dark-hover-color !default;
$theme-site-navbar-dark-bg-color:       $dark !default;

// Light
$theme-site-navbar-light-color:         rgba($black, .75) !default;
$theme-site-navbar-light-hover-color:   $black !default;
$theme-site-navbar-light-active-color:  $theme-site-navbar-light-hover-color !default;

// Lighter
$theme-site-navbar-lighter-color:         $black !default;
$theme-site-navbar-lighter-hover-color:   $primary !default;
$theme-site-navbar-lighter-active-color:  $theme-site-navbar-lighter-hover-color !default;
$theme-site-navbar-lighter-bg-color:      $primary !default;

// Fonts
$theme-site-navbar-primary-font-size:   16px !default;
$theme-site-navbar-secondary-font-size: 14px !default;

// Decorations
// Notes: 
// Primary/secondary both utilize the same settings currently
// Best to combine them into a single pull
$theme-site-navbar-link-decoration:         none !default;
$theme-site-navbar-link-hover-decoration:   underline !default;
$theme-site-navbar-link-active-decoration:  $theme-site-navbar-link-hover-decoration !default;

// Themes
$theme-site-navbar-map: (
  dark: (
    default:    $theme-site-navbar-dark-color,
    hover:      $theme-site-navbar-dark-hover-color,
    active:     $theme-site-navbar-dark-active-color,
    background: $theme-site-navbar-dark-bg-color
  ),
  light: (
    default:    $theme-site-navbar-light-color,
    hover:      $theme-site-navbar-light-hover-color,
    active:     $theme-site-navbar-light-active-color
  ),
  lighter: (
    default:    $theme-site-navbar-lighter-color,
    hover:      $theme-site-navbar-lighter-hover-color,
    active:     $theme-site-navbar-lighter-active-color,
    background: $theme-site-navbar-lighter-bg-color
  ),
  fonts: (
    primary:    $theme-site-navbar-primary-font-size,
    secondary:  $theme-site-navbar-secondary-font-size
  ),
  decorations: (
    default:    $theme-site-navbar-link-decoration,
    hover:      $theme-site-navbar-link-hover-decoration,
    active:     $theme-site-navbar-link-active-decoration
  )
);
```

Alixers505 commented on packages/themes/default/styles/variables/_navbar.scss:
```diff
@@ -0,0 +1,115 @@
+//  Spacing
+$theme-site-navbar-padding-x:           16px !default;
+$theme-site-navbar-padding-y:           0 !default;
+$theme-site-navbar-brand-margin-y: 0 !default;
```
Coolio. I didn't know this was a standard. I noticed this wasn't the case with a number of other assignments throughout some of the changed files, so I updated the standard variable file and the mixin file as well.

Alixers505 commented:
# Workflow Change Clarification for Overall Refactor

### Previous Flow for Navbar Link Colors, with code examples:

Step 1) Variables Assigned
`$theme-site-navbar-dark-color:  rgba($white, .75) !default;`

Step 2) Map Definition
- Has two maps, 1 for light and 1 for dark
```
$theme-site-navbar-dark-colors: (
     default: $theme-site-navbar-dark-color,
     etc...
);
```

Step 3) Function for Palette Switch
```
@function navbar-link-color($navbar-type, $color-type) {
    @if $navbar-type === dark {
        @return map-get($theme-site-navbar-dark-colors, $color-type);
    }
    @return map-get($theme-site-navbar-light-colors, $color-type);
}
```

Step 4) Section Specific Variable Assignments
```
$theme-site-navbar-primary-link-(etc)
$theme-site-navbar-secondary-link-(etc)
$theme-site-navbar-tertiary-link-(etc)
```
Note: This step is unnecessary with the current utilization, as every section was assigned to the exact same variables (ie: dark palette)

Also requires utilization of variable assignment $theme-site-navbar-secondary-type: dark !default, which has the same assignment as $theme-site-navbar-primary-type.

Step 5) Section Specific Color Map Definition
```
$theme-site-navbar-tertiary-colors: (
    default: $theme-site-navbar-tertiary-link-color,
    etc...
);
```
Note: This step is unnecessary, for the same reason as Step 4

Step 6) Mixin Definition
```
@mixin theme-navbar-link-color($color-map) {
    $color: map-get($color-map, default);
    etc...

    color: $color;
    svg {
        fill: $color;
    }
    etc...
};
```

Step 7) Mixin Utilization within navbar theme
` @include theme-navbar-link-color($theme-site-navbar-tertiary-colors);`

### New Flow for Navbar Link Color
Step 1) Same as above

Step 2) Map Definition
- Combined the dark and light color maps into a single nested color map
- Added lighter color palette for brands like bizbash, which utilize brand colors in the navbar
- Brands can override the map values the same way they could before, by simply reassigning a value to the variable name.
```
$theme-site-navbar-map: (
  dark: (
    default:    $theme-site-navbar-dark-color,
    hover:      $theme-site-navbar-dark-hover-color,
    active:     $theme-site-navbar-dark-active-color,
  ),
  light: (
    default:    $theme-site-navbar-light-color,
    hover:      $theme-site-navbar-light-hover-color,
    active:     $theme-site-navbar-light-active-color
  ),
  lighter: (
    default:    $theme-site-navbar-lighter-color,
    hover:      $theme-site-navbar-lighter-hover-color,
    active:     $theme-site-navbar-lighter-active-color,
  )
);
```

Step 3: Function
- Removed the function specific to switching between a dark and light map
- Added a function for nested map navigation across multiple needs (ie: it can be used for more than just color assignment)
```
@function map-deep-get($map, $keys...) {
  @each $key in $keys {
    $map: map-get($map, $key);
  }
  @return $map;
}
```

Removed Steps 4 and 5 from previous workflow, as they were redundant

Step 6 is now Step 4: Mixin 
- Updated mixin to utilize new function
- Switch requires adding a second parameter to denote which nested map you are referencing (dark, light, lighter, etc)

Step 7 is now Step 5: Mixin Utilization
- Same as before, except now you can define whether you are using dark, light, or lighter right in the mixin

### Benefits
- Saves 2 steps in the workflow
- Can now use the same workflow for decorations, background colors, etc (not just link colors)
-- To be implemented in the next stages of this phase
- Reduces the number of variables in the file by removing redundant assignments
- Starts the modularization of an extremely large variable file

Alixers505 commented:
Note: Will not start on Stage 2 of the Navbar Refactor until Stage 1 has been approved and merged, to prevent the risk of conflicts.

Alixers505 commented:
@zarathustra323 and @solocommand -- just checking in on this, since I haven't heard anything in about a week. I'm already seeing a conflict in the changes, due to updates to the bizbash design file. Going to resolve those, but I'd like to get this moving forward before we get any more conflicts.

Alixers505 commented:
Noticed that there are now more conflicts. I will wait to resolve these conflicts until y'all are ready to move forward with this, as I'm sure more will pop up....
